### PR TITLE
Extended Mat::reshape method to accept n-dimensional shapes.

### PR DESCRIPTION
**Pull request by ludv1x**

### CODE REVIEW

File: modules/core/src/matrix.cpp [hunk 1]
```diff
@@ -4257,6 +4257,64 @@ Mat Mat::reshape(int _cn, int _newndims, const int* _newsz) const
             return reshape(_cn, _newsz[0]);
     }
 
+    if (_newsz && _newndims > 0 && isContinuous())
+    {
+        CV_Assert(_cn >= 0);
+
+        size_t total_elem1_ref = this->total() * this->channels();
+        size_t total_elem1 = (_cn > 0) ? _cn : 1;
```
Comment (alalek) [inline modules/core/src/matrix.cpp]: Use same parameter behaviour:
`1` --> `this->channels()`
https://github.com/Itseez/opencv/blob/47e7a7a7ce11884ddaf46b0f7aa7774ee7708915/modules/core/src/matrix.cpp#L970-L971

+fix/add test for this case

File: modules/core/src/matrix.cpp [hunk 2]
```diff
@@ -4257,6 +4257,64 @@ Mat Mat::reshape(int _cn, int _newndims, const int* _newsz) const
...
+            }
+            else
+            {
+                _cn = auto_dim_size;
+            }
+        }
+        else
+        {
+            if (total_elem1 != total_elem1_ref)
+                CV_Error(CV_StsBadSize, "Requested matrix have differennt total matrix elements");
```
Comment (alalek) [inline modules/core/src/matrix.cpp]: typo in `different`
`has` or `matrices`?

**Comment by Nerei** [comment]
:+1: @vpisarev Vadim, could you take a look? DNN module depens on it. For me it's LGTM.

**Comment by alalek** [comment]
@ludv1x Please squash commits into one.

**Comment by alalek** [comment]
Try this:
```
git checkout -B master 0dc4074fad64753fad258ab395de8a3304446c27
git fetch https://github.com/Itseez/opencv.git master
git rebase -i FETCH_HEAD
# pick first commit, f/s for others
git push -f origin master
```

**Comment by ludv1x** [comment]
@alalek thanks)
I made similar first time. But a marking already merged commits as squashed was bad idea-)

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**